### PR TITLE
PK: ECDSA signing

### LINKED ISSUE #5274: PK: ECDSA signing
In `library/pk_wrap.c`, provide an implementation of `ecdsa_sign_wrap` to use `psa_sign_hash()` instead of `mbedtls_ecdsa_write_signature()`.

This will require creating a temporary PSA private key with appropriate permissions. An example of something similar, but with a public key, can be found in `ecdsa_verify_wrap()` in the same file (note: there are two definitions of `ecdsa_very_wrap()`, we want to look at the one used with `MBEDTLS_USE_PSA_CRYPTO`).

This will also require transcoding the resulting signature, as PSA doesn't output ECDSA signatures in the format we expect. We already have a helper function for that: `pk_ecdsa_sig_asn1_from_psa()`, used in `psa_sign_hash()`, which can be re-used here.

Similar: #5160 and #5162 are also creating a temporary PSA private key.

### PR DESCRIPTION
**Pull request by superna9999**
## Description
In `library/pk_wrap.c`, provide an implementation of `ecdsa_sign_wrap` to use `psa_sign_hash()` instead of `mbedtls_ecdsa_write_signature()`.

Resolves #5274

## Status
**READY**

## Requires Backporting
NO 

## Migrations
NO

## Additional comments
X509write suite failed:
```
Certificate Request check Server5 ECDSA, key_usage ................ FAILED
  memcmp( buf, check_buf, pem_len - 1 ) == 0
  at line 135, /home/narmstrong/projects/nordic/mbedtls/tests/suites/test_suite_x509write.function
```
because reference vector is calculated with `mbedtls_test_rnd_pseudo_rand`.

Using PSA DETERMINISTIC ECDSA in `pk_wrap` makes the test PASS, the only viable solution
is to disable the reference check and verify the signature instead.

## Todos
- [X] Implementation
- [x] Tests

## Steps to test or reproduce
test_suite_x509write and test_suite_pk should run clean

### CODE REVIEW

**Comment by superna9999** [comment]
The `X509write` failure seems to be because the csr checked in `x509_csr_check()` depends on `mbedtls_test_rnd_pseudo_rand` being used.

The probable solution would be to disable this test when `MBEDTLS_USE_PSA_CRYPTO` ?

File: tests/suites/test_suite_x509write.function
Comment (gilles-peskine-arm) [inline tests/suites/test_suite_x509write.function:96]: Please use `USE_PSA_INIT( )`, and `USE_PSA_DONE( )` in the cleanup. They'll do the right thing to initialize PSA crypto if needed and check that resources are properly deallocated.

**Comment by gilles-peskine-arm** [comment]
> The X509write failure seems to be because the csr checked in x509_csr_check() depends on mbedtls_test_rnd_pseudo_rand being used.
> The probable solution would be to disable this test when MBEDTLS_USE_PSA_CRYPTO ?

Ideally the dependencies on RNG usage should be minimized. I haven't looked closely at the code, would it be possible to run the test case but skip one specific test assertion? In any case, please add a comment explaining why the test case or part of it is skipped.

**Comment by superna9999** [comment]
@gilles-peskine-arm I fixed the case by calling `x509_crt_verifycsr()` when `MBEDTLS_USE_PSA_CRYPTO`, keeping the test case active but avoiding to use the reference file.

**Comment by superna9999** [comment]
@gilles-peskine-arm I don't have access to the failing checks, could you help figure out what's wrong ?

File: library/pk_wrap.c
Comment (gilles-peskine-arm) [inline library/pk_wrap.c:681]: MSVC complains:
```
     2>..\..\library\pk_wrap.c(681): warning C4267: '=' : conversion from 'size_t' to 'int', possible loss of data [C:\builds\workspace\mbed-tls-pr-head_PR-5518-head\worktrees\tmpdbrgm5pj\visualc\VS2010\mbedTLS.vcxproj]
```

Normally a size should be `size_t`. We sometimes have to convert sizes to `int` for compatibility with legacy or third-party APIs. In such cases, an explicit cast is often needed, and of course we need to be careful in case the size could overflow (which here, it looks like it can, with invalid input). But here I don't see why `key_len` is an `int *` rather than a `size_t *`.

Comment (superna9999) [reply library/pk_wrap.c:681]: @gilles-peskine-arm thanks for the feedback, I'll fix this and use a `size_t *`

File: library/pk_wrap.c
Comment (mprse) [inline library/pk_wrap.c:752]: Please handle status of the `psa_destroy_key`.

**Review by mprse — CHANGES_REQUESTED** [review]
Few minor issues mostly similar to other related PRs.

**Comment by superna9999** [comment]
I rebased against `development` after merge of #5581 and dropped the reverted commits moved to #5581.

**Comment by gilles-peskine-arm** [comment]
Sorry, I can't commit to reviewing this right now. I'll have limited review bandwidth in March.

File: library/pk_wrap.c
Comment (mprse) [reply library/pk_wrap.c:839]: Seems that this comment is not addressed.

Comment (superna9999) [reply library/pk_wrap.c:839]: @mprse indeed, I added the check on `ecdsa_verify_wrap()` instead, will add it to `ecdsa_sign_wrap()`

**Comment by mprse** [comment]
CI errors:

```
1.

[ 11%] Building C object library/CMakeFiles/mbedcrypto.dir/pk_wrap.c.o
/home/travis/build/ARMmbed/mbedtls/library/pk_wrap.c: In function ‘ecdsa_verify_wrap’:
/home/travis/build/ARMmbed/mbedtls/library/pk_wrap.c:721:15: error: implicit declaration of function ‘mbedtls_psa_err_translate_pk’ [-Werror=implicit-function-declaration]
         ret = mbedtls_psa_err_translate_pk( status );
               ^
[ 11%] Building C object library/CMakeFiles/mbedcrypto.dir/pkcs12.c.o
[ 11%] Building C object library/CMakeFiles/mbedcrypto.dir/pkcs5.c.o
[ 11%] Building C object library/CMakeFiles/mbedcrypto.dir/pkparse.c.o
cc1: all warnings being treated as errors
library/CMakeFiles/mbedcrypto.dir/build.make:569: recipe for target 'library/CMakeFiles/mbedcrypto.dir/pk_wrap.c.o' failed
make[2]: *** [library/CMakeFiles/mbedcrypto.dir/pk_wrap.c.o] Error 1
make[2]: *** Waiting for unfinished jobs....
CMakeFiles/Makefile2:405: recipe for target 'library/CMakeFiles/mbedcrypto.dir/all' failed
make[1]: *** [library/CMakeFiles/mbedcrypto.dir/all] Error 2
Makefile:140: recipe for target 'all' failed
make: *** [all] Error 2
Failed to build: config-suite-b.h
^^^^test_ref_configs: test/build: ref-configs (ASan build): tests/scripts/test-ref-configs.pl -> 1^^^^
******************************************************************
* Done, cleaning up 
* Thu Mar  3 13:45:18 UTC 2022
******************************************************************
!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!
FAILED: 1 components
test_ref_configs: test/build: ref-configs (ASan build): tests/scripts/test-ref-configs.pl -> 1
!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!
The command "tests/scripts/all.sh -k test_ref_configs" exited with 1.
10.32s$ tests/scripts/all.sh -k build_arm_linux_gnueabi_gcc_arm5vte build_arm_none_eabi_gcc_m0plus

2.   
CC    pkcs12.c
pk_wrap.c: In function ‘ecdsa_verify_wrap’:
pk_wrap.c:721:15: error: implicit declaration of function ‘mbedtls_psa_err_translate_pk’ [-Werror=implicit-function-declaration]
         ret = mbedtls_psa_err_translate_pk( status );
               ^
  CC    pkcs5.c
cc1: all warnings being treated as errors
Makefile:287: recipe for target 'pk_wrap.o' failed
make[1]: *** [pk_wrap.o] Error 1
make[1]: *** Waiting for unfinished jobs....
Makefile:18: recipe for target 'lib' failed
make: *** [lib] Error 2
^^^^build_arm_linux_gnueabi_gcc_arm5vte: build: arm-linux-gnueabi-gcc -march=arm5vte: make CC="${ARM_LINUX_GNUEABI_GCC_PREFIX}gcc" AR="${ARM_LINUX_GNUEABI_GCC_PREFIX}ar" CFLAGS='-Werror -Wall -Wextra -march=armv5te -O1' LDFLAGS='-march=armv5te' -> 2^^^^

3.
  CC    pk_wrap.c
pk_wrap.c: In function 'ecdsa_verify_wrap':
pk_wrap.c:721:9: error: implicit declaration of function 'mbedtls_psa_err_translate_pk' [-Werror=implicit-function-declaration]
         ret = mbedtls_psa_err_translate_pk( status );
         ^
  CC    pkcs12.c
  CC    pkcs5.c
cc1: all warnings being treated as errors
Makefile:287: recipe for target 'pk_wrap.o' failed
make[1]: *** [pk_wrap.o] Error 1
make[1]: *** Waiting for unfinished jobs....
Makefile:18: recipe for target 'lib' failed
make: *** [lib] Error 2
^^^^build_arm_none_eabi_gcc_m0plus: build: arm-none-eabi-gcc -mthumb -mcpu=cortex-m0plus: make CC="${ARM_NONE_EABI_GCC_PREFIX}gcc" AR="${ARM_NONE_EABI_GCC_PREFIX}ar" LD="${ARM_NONE_EABI_GCC_PREFIX}ld" CFLAGS='-std=c99 -Werror -Wall -Wextra -mthumb -mcpu=cortex-m0plus -Os' lib -> 2^^^^
```

**Comment by superna9999** [comment]
I rebased against `development` since merge of #5529 and fixed usage of new PSA to mbedtls PK error mapping functions

**Review by mprse — DISMISSED** [review]
All comments are addressed. CI is passing.
Approved.

File: library/pk_wrap.c
Comment (mpg) [inline library/pk_wrap.c:883]: Note: copy-pasta in the commit message. Can you rewrite history to fix it?

File: library/pk_wrap.c
Comment (mpg) [inline library/pk_wrap.c]: Can you add some documentation for this function? Something similar to what's done for `pk_ecdsa_sig_asn1_from_psa()`.

**Comment by mprse** [comment]
I think we need to restart CI here. The easiest way I think would be to amend last commit (without changing) and force-push.

**Comment by mpg** [comment]
> I think we need to restart CI here. The easiest way I think would be to amend last commit (without changing) and force-push.

In general yes. But it happens I'm reviewing and I'll request changes, so an alternative that works in this particular case it to wait until I complete my review :)

File: tests/suites/test_suite_x509write.function
Comment (mpg) [inline tests/suites/test_suite_x509write.function:126]: Alternatively, we could just wrap the corresponding declarations with an `#if` - except of course for `cert_req_check_file` which is a parameter.

**Review by mpg — CHANGES_REQUESTED** [review]
Looks pretty good. Just requesting documentation for the new internal function; the other two notes are optional.

File: library/pk_wrap.c
Comment (superna9999) [reply library/pk_wrap.c:883]: Yep will rewrite

**Review by mpg — APPROVED** [review]
LGTM, thanks for addressing my feedback!

**Review by mprse — COMMENTED** [review]
Delta review ok. 
LGTM.

**Review by mprse — APPROVED** [review]

**Pull request merged**
